yt/utilities/io_handler.py

```diff
@@ -2,10 +2,11 @@
 from collections import defaultdict
 from contextlib import contextmanager
 from functools import _make_key, lru_cache
-from typing import Tuple
+from typing import DefaultDict, Iterator, List, Mapping, Tuple
```

Comment:
I have no experience with using `Iterator` as a return type, and I can't find an existing example. Since the only place you're using this annotation is in a `NotImplemented` method, I'm not 100% sure I can trust mypy to reveal if you're using it wrong. Can you provide some additional insight ?

Also I note that importing it from typing is deprecated as of Python 3.9 (https://docs.python.org/3/library/typing.html#typing.Iterator) (this is also true for `Mapping`) so here's how to work around deprecation warnings:

```suggestion
import sys
from typing import DefaultDict, List, Tuple

if sys.version_info >= (3, 9):
    from collections.abc import Iterator, Mapping
else:
    from typing import Iterator, Mapping
```
further note that this -admittedly- bloated structure will be automatically cleaned up in the future when we require Python 3.9 thanks to pyupgrade, into:
```python
from collections.abc import Iterator, Mapping
```
 (see https://github.com/asottile/pyupgrade#python2-and-old-python3x-blocks)

Reply:
Well,  `_read_particle_coords` is a generator, so I was (somewhat blindly) just following the description in [`typing.Generator'](https://docs.python.org/3.6/library/typing.html#typing.Generator) for a `Generator` that only yields: you can either annotate with `Generator` using `SendType` and `ReturnType` set to `None` or just annotate with an `Iterator`, which is the route I went because it seemed clearer.

The main point is that we would want the object being yielded to resolve correctly. e.g., in a toy case using `Generator`:

```python
from typing import Generator, Tuple

def genthing() -> Generator[Tuple[int, int], None, None]:
    for thing in range(10):
        yield (thing, thing)

g = genthing()
print(type(g))  # <class 'generator'>
for gval in g:
    print(gval)
```
the hints correctly evaluate `g` as a `Generator` object and gval as `Tuple[int,int]`. e.g., in vscode:

![Screenshot from 2022-01-11 12-15-26](https://user-images.githubusercontent.com/22038879/148998755-7153f930-3d53-4b3f-b00a-642bdeac5f7e.png)

Happy to use a `Generator` instead of `Iterator` if that's any clearer. 

As to whether mypy is actually testing this here... I'm not sure either. The child classes that implement this function inherit the typing characteristics though, don't they? They do at least evaluate in vscode and pycharm correctly, but I don't know enough about mypy to know if that actually helps. I can say that during my attempts to type hint when the SPH frontends were overriding `_read_particle_coords` to additionally yield the `hsml` variable I got a lot of confusing mypy errors related to "number of values to unpack" errors where `_read_particle_coords` is called, and those errors were fixed by normalizing all the `_read_particle_coords` methods to what they are now. 

I also did wonder if making `BaseIOHandler` a true abstract base class would help? But I wasn't sure about how much of a refactor that would lead to so held off on the attempt.

Reply:
> Well, _read_particle_coords is a generator, so I was (somewhat blindly) just following the description in https://docs.python.org/3.6/library/typing.html#typing.Generator for a Generator that only yields

Thank you very much, I hadn't thought of looking there, and indeed this suits your use case.

> Happy to use a Generator instead of Iterator if that's any clearer.

no need, I was just uneducated. I agree now that `Iterator` is the appropriate annotation

> As to whether mypy is actually testing this here... I'm not sure either. The child classes that implement this function inherit the typing characteristics though, don't they?

Really not sure about that. Mypy doesn't seem to detect type inconsistencies that well through inheritance, see for instance

```python
from collections.abc import Iterator


class A:
    def foo(self) -> Iterator[int]:
        raise NotImplementedError

class B(A):
    def foo(self):
        for i in range(2):
            yield None # mypy doesn't detect the inconsistency here
```

that said, if I annotate the child method, mypy correctly errors in all cases that break the substitution principle. So annotating the base class is apparently not sufficient to maximise correctness, but it's a crucial first step (and we don't need to take this any further for the present PR)

> I also did wonder if making BaseIOHandler a true abstract base class would help? But I wasn't sure about how much of a refactor that would lead to so held off on the attempt.

I tried this approach on a toy example and it doesn't seem to make any difference as far as mypy is concerned. That said it would still make sense to do it because the class is _effectively_ abstract, so I think it's worth marking it as such, but that should happen in a different PR.

So in conclusion, nothing to do here, except maybe taking my suggestion above regarding imports

Reply:
Great! Will definitely put in the import suggestions.